tensorflow/compiler/mlir/xla/transforms/cycle_detector.cc

```diff
@@ -0,0 +1,361 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/mlir/xla/transforms/cycle_detector.h"
+
+#include <algorithm>
+#include "llvm/ADT/DenseSet.h"
+
+namespace mlir {
+
+namespace {
+
+using NodeSet = llvm::DenseSet<int32_t>;
+using OrderedNodeSet = OrderedSet<int32_t>;
+
+template <typename T>
+struct VecStruct {
+  typedef llvm::SmallVector<T, 4> type;
+};
+template <typename T>
+using Vec = typename VecStruct<T>::type;
+
+struct Node {
+  // rank number assigned by Pearce-Kelly algorithm
+  int32_t rank;
+  // Temporary marker used by depth-first-search
+  bool visited;
+  // User-supplied data
+  void* data;
+  // List of immediate predecessor nodes in graph
+  OrderedNodeSet in;
+  // List of immediate successor nodes in graph
+  OrderedNodeSet out;
+};
+
+}  // namespace
+
+
+struct GraphCycles::Rep {
+  Vec<Node*> nodes_;
```

Comment:
From clang-tidy check: the name of the struct member variable 'nodes_' violates the Style Guide; use lower case letters separated by '_' characters with no '_' character at the end, e.g. a_struct_data_member